Mathlib/CategoryTheory/Sites/Preserves.lean

```diff
@@ -0,0 +1,254 @@
+/-
+Copyright (c) 2023 Dagur Asgeirsson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Dagur Asgeirsson
+-/
+import Mathlib.CategoryTheory.Limits.Opposites
+import Mathlib.CategoryTheory.Limits.Preserves.Shapes.Products
+import Mathlib.CategoryTheory.Sites.SheafOfTypes
+import Mathlib.Tactic.ApplyFun
+
+/-!
+# Sheaves preserve products
+
+We prove that a presheaf which satisfies the sheaf condition with respect to certain presieves
+preserve "the corresponding products".
+
+More precisely, given a presheaf `F : Cᵒᵖ ⥤ Type*`, we have:
+
+## Main results
+
+* If `F` satisfies the sheaf condition with respect to the empty sieve on the initial object of `C`,
+  then `F` preserves terminal objects.
+See `preservesTerminalOfIsSheafForEmpty`.
+
+* If `F` furthermore satisfies the sheaf condition with respect to the presieve consisting of the
+  inclusion arrows in a coproduct in `C`, then `F` preserves the corresponding product.
+See `preservesProductOfIsSheafFor`.
+-/
+
+universe v u
+
+namespace CategoryTheory.Presieve
+
+open Limits Opposite
+
+variable {C : Type u} [Category.{v} C] (F : Cᵒᵖ ⥤ Type (max u v)) [HasInitial C]
+    (hF : (ofArrows (X := ⊥_ C) Empty.elim instIsEmptyEmpty.elim).IsSheafFor F)
+
+/--
+If `F` is a presheaf which satisfies the sheaf condition with respect to the empty presieve on the
+initial object, then `F` takes the initial object to the terminal object.
+-/
+noncomputable
+def isTerminal_obj_initial_of_isSheafFor_empty_presieve : IsTerminal (F.obj (op (⊥_ C))) := by
```

Comment:
This (and `hF`) should be stated in terms of the empty presieve on an initial object, and similar below